File: web/lib/types/workout-program.ts
// Wire types for Workout Programs (IMPL-15). Import-safe from client and
// server. Backend wire format uses uppercase Java enum names.

import type { BlockType, DemoFrame } from "./exercise";

export type WeekDay = "MON" | "TUE" | "WED" | "THU" | "FRI" | "SAT" | "SUN";

export type ProgramStatus = "DRAFT" | "ACTIVE" | "COMPLETED" | "ARCHIVED";
export type ProgramSource = "MANUAL" | "AI_GENERATED" | "AI_ASSISTED";
export type PhaseStatus = "LOCKED" | "ACTIVE" | "COMPLETED";
export type ScheduledStatus = "PLANNED" | "COMPLETED" | "SKIPPED";

export type IntensityKind = "RPE" | "PERCENT_1RM" | "NONE";

export type Intensity = {
  kind: IntensityKind;
  value: number | null;
};

export type DeloadModifier = {
  setsMultiplier: number | null;
  intensityDelta: number | null;
};

// Compact, read-only exercise summary embedded on deep/calendar responses so
// clients render a session without an N+1 fetch per prescription.
export type PrescriptionExercise = {
  exerciseId: string;
  name: string;
  primaryMuscles: string[];
  formCues: string[];
  demoFrames: DemoFrame[];
};

// One set as actually performed. Populated on completed-session snapshots
// (history import + the IMPL-17 logger); null/absent on plan templates. Reps
// are null in the imported export — only the weight was recorded. RPE, rest,
// and the per-set timestamp are full actuals from live logging (ADR-0012
// Decision 2) — always nullable, never required.
export type LoggedSet = {
  weightLbs: number | null;
  reps: number | null;
  rpe: number | null;
  restSeconds: number | null;
  completedAt: string | null;
};

export type Prescription = {
  exerciseId: string;
  orderIndex: number;
  sets: number | null;
  repsMin: number | null;
  repsMax: number | null;
  durationSeconds: number | null;
  intensity: Intensity | null;
  restSeconds: number | null;
  tempo: string | null;
  notes: string | null;
  deloadModifier: DeloadModifier | null;
  loggedSets: LoggedSet[] | null;
  exercise: PrescriptionExercise | null;
  // Inline validation note set by the backend validator (e.g. "not executable
  // at this gym"). Used to flag the offending row.
  validationError?: string | null;
};

export type Block = {
  blockId: string;
  type: BlockType;
  title: string;
  orderIndex: number;
  prescriptions: Prescription[];
};

export type WorkoutDay = {
  dayId: string;
  label: string;
  dayOfWeek: WeekDay;
  locationId: string;
  locationName: string;
  orderIndex: number;
  blocks: Block[];
};

export type Phase = {
  phaseId: string;
  title: string;
  focus: string;
  orderIndex: number;
  status: PhaseStatus;
  weeks: number;
  deloadWeekIndex: number | null;
  targetStartDate: string;
  targetEndDate: string;
  days: WorkoutDay[];
};

// Shallow list response.
export type WorkoutProgramResponse = {
  programId: string;
  title: string;
  description: string;
  goalId: string | null;
  status: ProgramStatus;
  source: ProgramSource;
  startDate: string;
  trainingDays: WeekDay[];
  totalWeeks: number;
  phaseCount: number;
  completedPhaseCount: number;
  createdAt: string;
  updatedAt: string;
};

// Deep response with phases → days → blocks → prescriptions.
export type WorkoutProgramDeepResponse = {
  programId: string;
  title: string;
  description: string;
  goalId: string | null;
  goalTitle: string | null;
  status: ProgramStatus;
  source: ProgramSource;
  startDate: string;
  trainingDays: WeekDay[];
  phases: Phase[];
  createdAt: string;
  updatedAt: string;
  completedAt: string | null;
};

// A DayResponse: same shape as a WorkoutDay's session, used by ScheduledWorkout.
export type DayResponse = WorkoutDay;

// Lightweight history counts for the Workouts hub (no full session payloads).
export type WorkoutHistorySummary = {
  count: number;
  lastWorkoutDate: string | null;
};

export type ScheduledWorkoutResponse = {
  scheduledId: string;
  date: string;
  // Owning program. `scheduledId` is only unique within a program (IMPL-17 D1),
  // so cross-program reads (Workout History) need this to address the
  // completion upsert. Optional: program-scoped responses may omit it.
  programId?: string | null;
  phaseId: string;
  dayId: string;
  dayLabel: string;
  weekIndexInPhase: number;
  isDeload: boolean;
  locationId: string;
  locationName: string;
  status: ScheduledStatus;
  session: DayResponse;
  // Performed-session metadata (history import). Null for PLANNED sessions.
  completedAt: string | null;
  durationSeconds: number | null;
};

// ── Session completion (IMPL-17 D2) ──────────────────────────────────
//
// PUT /api/me/workout-programs/{programId}/sessions/{scheduledId} — the
// idempotent "log result / edit actuals" upsert from ADR-0012. Prescriptions
// have no id, so logged sets key by (blockId, orderIndex) against the session
// snapshot; unknown keys are a 400. Repeat PUTs replace actuals and re-run the
// backend fan-out.

export type LoggedSetInput = {
  weightLbs: number | null;
  reps: number | null;
  rpe: number | null;
  restSeconds: number | null;
  completedAt: string | null;
};

export type LoggedPrescriptionInput = {
  blockId: string;
  orderIndex: number;
  sets: LoggedSetInput[];
};

export type CompleteSessionRequest = {
  status: "COMPLETED" | "SKIPPED";
  completedAt: string | null; // ISO instant; required when COMPLETED
  durationSeconds: number | null; // required when COMPLETED
  logged: LoggedPrescriptionInput[]; // SKIPPED clears actuals (send [])
};

// ── Create / update request shapes (mirror backend records) ──────────

export type PrescriptionInput = {
  exerciseId: string;
  sets: number | null;
  repsMin: number | null;
  repsMax: number | null;
  durationSeconds: number | null;
  intensity: Intensity | null;
  restSeconds: number | null;
  tempo: string | null;
  notes: string | null;
  deloadModifier: DeloadModifier | null;
};

export type BlockInput = {
  type: BlockType;
  title: string;
  prescriptions: PrescriptionInput[];
};

export type DayInput = {
  label: string;
  dayOfWeek: WeekDay;
  locationId: string;
  blocks: BlockInput[];
};

export type PhaseInput = {
  title: string;
  focus: string;
  weeks: number;
  deloadWeekIndex: number | null;
  days: DayInput[];
};

export type ScheduleInput = {
  trainingDays: WeekDay[];
  dayLocations: Partial<Record<WeekDay, string>>;
};

export type CreateProgramRequest = {
  title: string;
  description: string;
  goalId: string | null;
  schedule: ScheduleInput;
  startDate: string;
  source: ProgramSource;
  phases: PhaseInput[];
};

export type UpdateProgramRequest = Partial<{
  title: string;
  description: string;
  goalId: string | null;
  schedule: ScheduleInput;
  startDate: string;
  status: ProgramStatus;
}>;

// ── Persistent chat threads (mirror backend ChatThreadResponse) ──────

// The schedule chosen in the pre-chat setup form. Fixed for the lifetime of a
// thread; the per-gym exercise allow-lists derive from these gyms.
export type WorkoutProgramChatSchedule = {
  trainingDays: WeekDay[];
  dayLocations: Partial<Record<WeekDay, string>>;
};

// GET /api/me/workout-programs/chat/threads element.
export type WorkoutProgramChatThread = {
  threadId: string;
  title: string;
  schedule: WorkoutProgramChatSchedule | null;
  goalId: string | null;
  createdAt: string;
  updatedAt: string;
};

export type WorkoutProgramChatMessageRole = "USER" | "ASSISTANT";

// GET /api/me/workout-programs/chat/{threadId} element. `proposalJson` is a
// JSON string of { program: WorkoutProgramDeepResponse, issues: string[] } on
// assistant turns that produced a proposal; null otherwise.
export type WorkoutProgramChatMessage = {
  messageId: string;
  role: WorkoutProgramChatMessageRole;
  content: string;
  proposalJson: string | null;
  createdAt: string;
};

// Parsed shape of an assistant message's `proposalJson` and the SSE `proposal`
// event payload.
export type WorkoutProgramProposalPayload = {
  program: WorkoutProgramDeepResponse;
  issues: string[];
};

export const WEEK_DAYS: WeekDay[] = [
  "MON",
  "TUE",
  "WED",
  "THU",
  "FRI",
  "SAT",
  "SUN",
];

export const WEEK_DAY_LABEL: Record<WeekDay, string> = {
  MON: "Mon",
  TUE: "Tue",
  WED: "Wed",
  THU: "Thu",
  FRI: "Fri",
  SAT: "Sat",
  SUN: "Sun",
};

File: android/core-domain/src/main/java/com/gte619n/healthfitness/domain/workouts/program/WorkoutProgram.kt
package com.gte619n.healthfitness.domain.workouts.program

import com.gte619n.healthfitness.domain.common.DayOfWeek
import java.time.Instant
import java.time.LocalDate

// Workout-program domain models (IMPL-AND-15, read-only). These mirror the
// IMPL-15 backend read contract (WorkoutProgramResponse shallow,
// WorkoutProgramDeepResponse, PhaseResponse, WorkoutDayResponse, BlockResponse,
// PrescriptionResponse, ScheduledWorkoutResponse) — adapted to typed
// LocalDate/Instant/DayOfWeek via the core-data Moshi adapters. No proposal /
// chat types. ADR-0012 adds the performed-session actuals ([LoggedSet],
// `ScheduledWorkout.completedAt`/`durationSeconds`); the live-logging draft
// types live in `domain.workouts.session`.

enum class ProgramStatus { DRAFT, ACTIVE, COMPLETED, ARCHIVED }

enum class ProgramSource { MANUAL, AI_GENERATED, AI_ASSISTED }

enum class ProgramPhaseStatus { LOCKED, ACTIVE, COMPLETED }

enum class BlockType { WARMUP, MOBILITY, CARDIO, MAIN, ACCESSORY, CORE, COOLDOWN, STRETCH }

enum class IntensityKind { RPE, PERCENT_1RM, NONE }

enum class ScheduledStatus { PLANNED, COMPLETED, SKIPPED }

data class Intensity(val kind: IntensityKind, val value: Double?)

data class DeloadModifier(val setsMultiplier: Double?, val intensityDelta: Double?)

/**
 * Non-binding per-phase (or program-level fallback) calorie/macro guidance
 * written alongside a designed program (IMPL-18 S3/R4). Display-only — the user
 * still logs food in the nutrition module. All fields nullable.
 */
data class NutritionGuidance(
    val kcal: Int? = null,
    val proteinG: Int? = null,
    val carbsG: Int? = null,
    val fatG: Int? = null,
    val note: String? = null,
) {
    /** True when every field is empty — treat as "no guidance" and omit from the UI. */
    val isEmpty: Boolean
        get() = kcal == null && proteinG == null && carbsG == null && fatG == null &&
            note.isNullOrBlank()
}

/**
 * A single demo still for an exercise. IMPL-19 replaced the fixed
 * START/MID/END triad with a per-exercise frame plan: [key]/[label]/[caption]
 * are denormalized from the plan's `FrameSpec`, [order] drives display order,
 * and [phase] is the deprecated legacy enum (nullable, read-only for old docs).
 */
data class DemoFrame(
    val key: String = "",
    val label: String = "",
    val caption: String = "",
    val order: Int = 0,
    val imageUrl: String?,
    val phase: String? = null,
)

/** Compact, embedded exercise info for rendering a prescription + its demo. */
data class ExerciseSummary(
    val exerciseId: String,
    val name: String,
    val primaryMuscles: List<String>,
    val formCues: List<String>,
    val demoFrames: List<DemoFrame>,
)

/**
 * One performed set's full actuals (ADR-0012 Decision 2 / IMPL-17 D3). Every
 * field is nullable: imported-history rows are weight-only (reps null) and the
 * logger keeps everything beyond weight/reps skippable.
 */
data class LoggedSet(
    val weightLbs: Double? = null,
    val reps: Int? = null,
    val rpe: Double? = null,
    val restSeconds: Int? = null,
    val completedAt: Instant? = null,
    /** Held time for a timed exercise (stretch/mobility/cardio); the time-based counterpart to [reps]. */
    val durationSeconds: Int? = null,
)

data class Prescription(
    val exerciseId: String,
    val orderIndex: Int,
    val sets: Int?,
    val repsMin: Int?,
    val repsMax: Int?,
    val durationSeconds: Int?,
    val intensity: Intensity?,
    val restSeconds: Int?,
    val tempo: String?,
    val notes: String?,
    val deloadModifier: DeloadModifier?,
    /** Embedded by the backend; null only if the backend omits it. */
    val exercise: ExerciseSummary?,
    /**
     * Actual sets performed (ADR-0012); populated only for completed/imported
     * sessions, empty until the session is logged.
     */
    val loggedSets: List<LoggedSet> = emptyList(),
    /** IMPL-18: concrete history-grounded prescribed load; null → fall back to [intensity]. */
    val targetWeightLbs: Double? = null,
    /** IMPL-18: short "why" for the prescribed load (e1RM / last done / ramp), shown on tap (R6). */
    val loadBasis: String? = null,
) {
    /**
     * A timed exercise (stretch / mobility / cardio hold) — logged by held time
     * rather than reps. True when a duration is prescribed and no rep target is.
     */
    val isTimed: Boolean
        get() = durationSeconds != null && repsMin == null && repsMax == null
}

data class Block(
    val blockId: String,
    val type: BlockType,
    val title: String,
    val orderIndex: Int,
    val prescriptions: List<Prescription>,
)

data class WorkoutDay(
    val dayId: String,
    val label: String,
    val dayOfWeek: DayOfWeek,
    val locationId: String,
    /** Resolved by the backend for display; may be null. */
    val locationName: String?,
    val orderIndex: Int,
    val blocks: List<Block>,
)

data class ProgramPhase(
    val phaseId: String,
    val title: String,
    val focus: String?,
    val orderIndex: Int,
    val status: ProgramPhaseStatus,
    val weeks: Int,
    val deloadWeekIndex: Int?,
    val targetStartDate: LocalDate?,
    val targetEndDate: LocalDate?,
    /** Empty in the shallow list response. */
    val days: List<WorkoutDay> = emptyList(),
    /** IMPL-18: per-phase calorie/macro guidance (display-only); null = none. */
    val nutritionGuidance: NutritionGuidance? = null,
)

data class WorkoutProgram(
    val programId: String,
    val title: String,
    val description: String?,
    val goalId: String?,
    /** Present on the deep response (so the detail can label the goal link). */
    val goalTitle: String? = null,
    val status: ProgramStatus,
    val source: ProgramSource,
    val startDate: LocalDate?,
    val trainingDays: List<DayOfWeek>,
    val createdAt: Instant,
    val updatedAt: Instant,
    val completedAt: Instant? = null,
    // Backend-supplied roll-ups on the shallow list (no client computation).
    val totalWeeks: Int = 0,
    val phaseCount: Int = 0,
    val completedPhaseCount: Int = 0,
    /** Empty in the shallow list; populated on the deep response. */
    val phases: List<ProgramPhase> = emptyList(),
    /** IMPL-18: program-level nutrition fallback when phases carry none; null = none. */
    val nutritionGuidance: NutritionGuidance? = null,
) {
    /** "completed of total" phase pair for progress rendering. */
    val phaseProgress: Pair<Int, Int>
        get() = if (phases.isNotEmpty()) {
            phases.count { it.status == ProgramPhaseStatus.COMPLETED } to phases.size
        } else {
            completedPhaseCount to phaseCount
        }
}

data class ScheduledWorkout(
    val scheduledId: String,
    val date: LocalDate,
    val phaseId: String,
    val dayId: String,
    val dayLabel: String,
    val weekIndexInPhase: Int,
    val isDeload: Boolean,
    val locationId: String,
    val locationName: String?,
    val status: ScheduledStatus,
    /** A full day object (same shape as a deep day); present on the calendar. */
    val session: WorkoutDay? = null,
    /** Outcome fields (ADR-0012); set once the session is COMPLETED. */
    val completedAt: Instant? = null,
    val durationSeconds: Int? = null,
    /**
     * Owning program + phase titles, resolved by the Workout History read so the
     * list can draw program/phase delineation headers. Null elsewhere (calendar).
     */
    val programTitle: String? = null,
    val phaseTitle: String? = null,
)

/**
 * One page of Workout History rows (newest first). [hasMore] tells the caller
 * whether a further [page] exists, driving the screen's load-on-scroll.
 */
data class WorkoutHistoryPage(
    val items: List<ScheduledWorkout>,
    val page: Int,
    val total: Int,
    val hasMore: Boolean,
)

/**
 * Activation failed validation (HTTP 422): the backend returned the actionable
 * issue list (same shape as the designer's commit 422). Carried as a typed
 * failure so the UI can surface the specific issues inline instead of a generic
 * "couldn't activate" (IMPL-STAB G1). [issues] is never empty.
 */
class ProgramActivationInvalidException(val issues: List<String>) :
    Exception(issues.joinToString("; "))

/**
 * Read-only repository for workout programs. Mirrors the IMPL-15 read surface;
 * no mutation, no chat, no logging in v1.
 */
interface WorkoutProgramRepository {
    /** Shallow list of the signed-in user's programs. */
    suspend fun list(): Result<List<WorkoutProgram>>

    /** Deep program (phases → days → blocks → prescriptions + embedded summaries). */
    suspend fun get(programId: String): Result<WorkoutProgram>

    /** Scheduled sessions in the inclusive date range, for the "this week" strip. */
    suspend fun calendar(
        programId: String,
        from: LocalDate,
        to: LocalDate,
    ): Result<List<ScheduledWorkout>>

    /**
     * One page of COMPLETED sessions across all programs, newest first, deep
     * enough to review (each carries its blocks → prescriptions → logged sets).
     * Paged for the history screen's load-on-scroll. Online-only, read-only.
     */
    suspend fun workoutHistoryPage(page: Int, size: Int): Result<WorkoutHistoryPage>

    /**
     * The program's effective nutrition guidance (active phase's, else
     * program-level), or null when it has none. Online-only.
     */
    suspend fun nutritionGuidance(programId: String): Result<NutritionGuidance?>

    /** Apply the program's guidance as the user's macro target; returns the saved macros. Online-only. */
    suspend fun applyNutritionTarget(programId: String): Result<com.gte619n.healthfitness.domain.nutrition.Macros>

    /**
     * Activate a program (materialize sessions + mark ACTIVE) and refresh the
     * local mirror so the detail/list reflect it. Returns the materialized
     * sessions. Online-only.
     */
    suspend fun activate(programId: String): Result<List<ScheduledWorkout>>

    /**
     * Edit a program's metadata (title/description) via PATCH and refresh the
     * mirror so the detail reflects it without waiting for a sync. Structural
     * edits go through the conversational designer, not this. Online-only
     * (IMPL-STAB G4). Returns the updated deep program.
     */
    suspend fun updateDetails(
        programId: String,
        title: String,
        description: String?,
    ): Result<WorkoutProgram>
}
